pkg/pillar/cmd/zedagent/handlevolume.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2017-2018 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package zedagent
+
+// volume specific parser/utility routines
+
+import (
+	"bytes"
+	"crypto/sha256"
+
+	zconfig "github.com/lf-edge/eve/api/go/config"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	"github.com/satori/go.uuid"
+	log "github.com/sirupsen/logrus"
+)
+
+var volumeHash []byte
+
+// volume parsing routine
+func parseVolumeConfig(ctx *getconfigContext,
+	config *zconfig.EdgeDevConfig) {
+
+	log.Debugf("Started parsing volume config")
+	cfgVolumeList := config.GetVolumes()
+	h := sha256.New()
+	for _, cfgVolume := range cfgVolumeList {
+		computeConfigElementSha(h, cfgVolume)
+	}
+	newHash := h.Sum(nil)
+	if bytes.Equal(newHash, volumeHash) {
+		return
+	}
+	log.Infof("parseVolumeConfig: Applying updated config "+
+		"Last Sha: % x, "+
+		"New  Sha: % x, "+
+		"Num of cfgContentInfo: %d",
+		volumeHash, newHash, len(cfgVolumeList))
+
+	volumeHash = newHash
+
+	// First look for deleted ones
+	items := ctx.pubVolumeConfig.GetAll()
+	for idStr := range items {
+		found := false
+		for _, cfgVolume := range cfgVolumeList {
+			if cfgVolume.GetUuid() == idStr {
+				found = true
+				break
+			}
+		}
+		// content tree not found, delete
+		if !found {
+			log.Infof("parseVolumeConfig: deleting %s\n", idStr)
+			unpublishVolumeConfig(ctx, idStr)
+		}
+	}
+
+	for _, cfgVolume := range cfgVolumeList {
+		volumeConfig := new(types.NewVolumeConfig)
+		volumeConfig.VolumeID, _ = uuid.FromString(cfgVolume.GetUuid())
+		volumeOrigin := cfgVolume.GetOrigin()
+		if volumeOrigin != nil {
+			switch volumeOrigin.GetType() {
+			case zconfig.VolumeContentOriginType_VCOT_DOWNLOAD:
+				volumeConfig.ContentID, _ = uuid.FromString(volumeOrigin.GetDownloadContentTreeID())
+			default:
```

Comment:
Shouldn't this be a check for BLANK, and the default should be an error log?
Or beter, pass the GetType in VolumeConfig so that volumemgr can pass back an error in VolumeStatus.

Reply:
Makes sense. I will change it accordingly. Thanks!